includes/class-wcs-failed-scheduled-action-manager.php

```diff
@@ -83,8 +83,8 @@ public function log_action_scheduler_failure( $action_id, $error ) {
 
 		$subscription_action = $this->get_action_hook_label( $action->get_hook() );
 
-		// Not log when there is not a hook associated with the scheduled action
-		if ( strpos( $error->getMessage(), 'not be executed as no callbacks are registered' ) ) {
+		// Don't log errors for Trial End and Payment Rety scheduled actions when there are no callbacks associated with the scheduled action.
+		if ( strpos( $error->getMessage(), 'not be executed as no callbacks are registered' ) && in_array( $action->get_hook(), [ 'woocommerce_scheduled_subscription_trial_end', 'woocommerce_scheduled_subscription_payment_retry' ], true ) ) {
```

Comment:
Is there a programmatic way to detect the "no callbacks" error? Ideally we wouldn't have to compare with an error message string fragment – fragile.

Reply:
@haszari Doesn't seem like there is as they just throw a standard exception. Here's a reference to the code in Action Scheduler: https://github.com/woocommerce/action-scheduler/blob/520790a8f770c900427e5bb350a2c8a4e86168e6/classes/actions/ActionScheduler_Action.php#L37

Reply:
We could add an additional check for `! has_action( $subscription )` to confirm there's no callbacks but we would still need to do a string comparison on the error message to know what error we're checking against

I agree it's fragile but not sure of other solutions.

Reply:
Thanks for checking @mattallan. Is it possible we're doing something wrong here? Or is this something we should feed back about – that exception looks harsh at face value, it's informational, there's no danger (no exception IMO!).

Reply:
> Is it possible we're doing something wrong here?

Do you mean: is Subscriptions using AS in an unintended way? Or is this change made to AS too harsh?

I don't think we're necessarily doing anything wrong in Subscriptions as the trial end period action is more for 3rd parties to hook onto i.e. the Memberships plugin or email services/plugins to send emails to customers

Another option to solve this would be for us to attach our own callback that just fires off a different hook name for the trial ending event that 3rd parties could use 🤔 

i.e. something like:

```
add_action( 'woocommerce_scheduled_subscription_trial_end', 'trigger_subscription_trial_end_event' );

function trigger_subscription_trial_end_event( $subscription_id ) {
    do_action( 'woocommerce_subscriptions_trial_end', $subscription_id );
}
```

Seems a bit silly to do this just to suppress the errors from AS but I'm open to ideas!!

> that exception looks harsh at face value, it's informational, there's no danger (no exception IMO!).

For more context, the exception is caught on Action Scheduler's side and then the error/exception is sent on the `action_scheduler_failed_execution` hook which we're hooked onto in Subscriptions to handle these error.

In the future there could be errors from AS that we want to ignore so it would be nice for AS to tell us the type of error via some key/ID so we don't have to check strings.

Reply:
It would be good to update the generic exception in Action Scheduler to a custom exception to allow for easier detection and handling. Something like `ActionScheduler_MissingHooksException`.

Then detection would be 
```php
if ( $error instanceof ActionScheduler_MissingHooksException ) { 
   ...
}
```

Reply:
However, we'd still have to check for the specific string for backwards compatibility.

Reply:
> Do you mean: is Subscriptions using AS in an unintended way? Or is this change made to AS too harsh?

That's exactly what I mean. Are we doing things in a way that AS doesn't expect (e.g. logging all exceptions) or does need to better understand our use case?

> In the future there could be errors from AS that we want to ignore so it would be nice for AS to tell us the type of error via some key/ID so we don't have to check strings.

That would work, but IMO would be overengineering. Ideally we wouldn't need to selectively ignore any exceptions – that sounds like a patch on a bad API. Maybe the exceptions should have levels – i.e. the ones that are causing pain or signaling a problem vs. others that are informational or "something maybe fishy here".

Your approach of using a custom hook sounds elegant to me. I also think it would be reasonable to leave this as is and educate merchants/3PDs on what it means – i.e. it's informational. 

Anyway, up to y'all how we handle this, but it sounds like we should chat with AS squad to make sure we're aligned. We shouldn't have to patch or work around core – we're on the same team. Maybe we could ask for a review from that team on this issue?

Reply:
I just noticed the error message from Action Scheduler is translatable, so we won't be able to use this approach that we're using in this PR anyway 😢 

So I think we go with my other approach which is adding a callback for the trial end hook inside WC Subscriptions to _silence_ these "no callback errors"

> Something like ActionScheduler_MissingHooksException.

Yeah that would be nice if it were introduced at the beginning when they added this ! has_action check, but now we'd need to support the generic and non-generic exception which would be kind of messy

Reply:
There's no particular urgency on solving this – so if we want an upstream fix, we can lobby for that and follow up later. 

Though @mattallan's idea sounds lean and elegant.